Check guardedness of fixpoints also in erasable subterms (#15434)
coq/coq: doc/changelog/01-kernel/15434-master+check-guard-also-erasable-subterms.rst

- **Changed:**
  Fixpoints are now expected to be guarded even in subterms erasable
  by reduction, thus getting rid from an artificial obstacle
  preventing to lift the assumption of weak normalization of Coq to an
  assumption of strong normalization; for instance (barring
  implementation bugs) termination of the type-checking algorithm of
  Coq is now restored (of course, as usual, up to the assumption of

>   the consistency of set theory and type theory, i.e., equivalently,

Reviewer: ```suggestion
  the consistency of set theory and type theory, or equivalently,
```

  up to the weak normalization of type theory, a "physical"
  assumption, which has not been contradicted for decades and which

>   specialists commonly believe to be a truth)

Reviewer: This whole description is a bit of a mouthful however. Perhaps @jfehrle can have a stab at it.

Reviewer 2: As a starting point, readers IMHO shouldn't be taught about weak or strong normalization in a side note from the changelog. What about:
```
- **Changed:**
  Fixpoints are now expected to be guarded even in subterms erasable
  by reduction. This fixes a breakage of strong normalization, and
  eliminates one potential source of infinite loops during type-checking.
```
I'm not sure this agrees on the fact with @herbelin, because I can't quite parse the original.

If there's background to be given, it could go in the manual proper, but surely other introductions exist.

Reviewer 3: A mouthful indeed.  How about something along these lines?  Maybe without the parenthetical comment?  And corrected if I've gotten the meaning wrong?

```
  Fixpoints must now be guarded even in subterms erasable
  by reduction.  This ensures termination of Coq’s type-checking
  algorithm (subject to the commonly believed assumptions that set theory
  and type theory are both consistent).
```

If there are no implementation errors, then is it still possible to get an infinite loop?

I have little idea what "up to the weak normalization of type theory, a 'physical' assumption" is supposed to mean but it seems like too much information.

FWIW, there's no definition for "strong normalization" in the doc and only 4 uses of the term.  And no mentions of "weak normalization" in the doc.

Author: I agree the first version was terrible. Here is a new possible iteration based on your suggestions:
```
  Fixpoints must now be guarded even in subterms erasable
  by reduction. This ensures termination of Coq's type-checking
  algorithm and termination of the various reduction strategies 
  available in Coq (subject to the commonly believed assumptions that       
  set theory and type theory are consistent).    
```
I removed the "both" because one of them is enough. I added reduction strategies since this is one of the easiest observable consequences. I renounced to "strong normalization" which is indeed technical.

Reviewer 3: That's a good start.  A couple thoughts:
- It sounds like this PR may cause some expressions that were previously accepted to no longer be accepted.  What is the impact on the user?  What should they do if they're affected?
- "subterms erasable by reduction" refers to the type checking process (and not to, say, `cbv`)?  What makes a subterm erasable by reduction?  In the "Core Language / Typing Rules" section of the doc, there's only a brief mention of "reduction" under "Subtyping Rules : Normal Form".
- The doc doesn't explain "guarded".  There are little bits under the `CoFixpoint` command, the `Guard Checking` and `Print Typing Flags` flags and the `bypass_check(guard)` attribute.  Should `Guard Checking` and `bypass_check(guard)` go away or remain available?

It would be nice to address the last two points somewhat better if you're willing, but it doesn't have to be perfect.  (And we still need to polish off #15489--I was on vacation last week.)

Author: > That's a good start. A couple thoughts:
> 
>     * It sounds like this PR may cause some expressions that were previously accepted to no longer be accepted.  What is the impact on the user?  What should they do if they're affected?

I suspect that it'll have basically no impact. On the whole CI, we found only one example which had to be changed. I hope that the change log message is clear enough so that, if ever it happens, the readers would connect the failure of their Fixpoint with the change and adapt their fixpoint to remove the non-termination in an erasable subterm (see below for the definition).

>     * "subterms erasable by reduction" refers to the type checking process (and not to, say, `cbv`)?  What makes a subterm erasable by reduction?  In the "Core Language / Typing Rules" section of the doc, there's only a brief mention of "reduction" under "Subtyping Rules : Normal Form".

By "subterms erasable by reduction" is meant a subterm that disappears in the normal form of the term (using `lazy` or `cbv` or any reduction to normal form). A typical example is `let x := 0 in true` where `0`, bound to a variable which does not occur, is an erasable subterm. The typical artificial but faulty example is precisely `Fixpoint f x := let _ := f x in 0.`

>     * The doc doesn't explain "guarded".  There are little bits under the `CoFixpoint` command, the `Guard Checking` and `Print Typing Flags` flags and the `bypass_check(guard)` attribute.  Should `Guard Checking` and `bypass_check(guard)` go away or remain available?

I believe that the terminology "guarded" is now quite established in the community (and it is actually more established now that when the manual was first written).

A possible improvement would be to replace in the Fixpoint section the paragraph "To be accepted, a Fixpoint definition has to satisfy syntactical constraints on a special argument called the decreasing argument." with "To be accepted, a Fixpoint definition has to be guarded. That is, it has to satisfy some syntactical constraints on a special argument called the decreasing argument."  What do you think?

> It would be nice to address the last two points somewhat better if you're willing, but it doesn't have to be perfect. 

Possibly, I can do the change proposed above to the doc.

> (And we still need to polish off #15489--I was on vacation last week.)

Sure!

Reviewer 3: > I believe that the terminology "guarded" is now quite established in the community (and it is actually more established now that when the manual was first written).

We should make it easy for newcomers to learn what guardedness is, rather than keeping it as the esoteric knowledge of the cognoscenti.  IMHO.

We can do the doc improvement in a separate PR.  If you wish to just state the facts in outline form, I can turn that into proper prose.
